File: backend/functions/agentRoutes.js
// 📁 functions/agentRoutes.js – Express Gen 2 + Sicurezza Trigger IA

const express = require("express");
const admin = require("firebase-admin");
const { onRequest } = require("firebase-functions/v2/https");
const { verifyToken } = require("../middlewares/verifyToken");
const withRateLimit = require("../middlewares/withRateLimit");

const { trackIAUsage } = require("./usageTracker");
const { sendNotification } = require("./lib/sendNotification");
const { generateDocumentTags } = require("./lib/generateDocumentTags");
const { matchEvent } = require("./lib/eventMatcher");

if (!admin.apps.length) admin.initializeApp();
const db = admin.firestore();
const router = express.Router();

// 🔐 Middleware globale
router.use(verifyToken);

// 📊 Logging globale
router.use((req, res, next) => {
  console.log(`📥 [${req.method}] ${req.originalUrl} – UID: ${req.user?.uid}`);
  next();
});

// 🛡️ Rate limiter personalizzato
const rateLimiter = withRateLimit(10, 60_000);

// --- INIZIO ROTTE ---
// ✅ Recupera tutte le azioni IA
router.get("/actions/:userId", async (req, res) => {
  try {
    const { userId } = req.params;
    const snapshot = await db
      .collection("ai_agent_hub")
      .doc(userId)
      .collection("actions")
      .orderBy("startedAt", "desc")
      .get();

    const actions = snapshot.docs.map((doc) => doc.data());
    res.status(200).json(actions);
  } catch (err) {
    console.error("❌ getAgentActions:", err);
    res.status(500).json({ error: "Errore recupero azioni IA" });
  }
});

// ✅ Elimina azione IA
router.delete("/actions/:userId/:actionId", async (req, res) => {
  try {
    const { userId, actionId } = req.params;
    await db
      .collection("ai_agent_hub")
      .doc(userId)
      .collection("actions")
      .doc(actionId)
      .delete();
    res.status(200).json({ message: "✅ Azione eliminata" });
  } catch (err) {
    console.error("❌ deleteAgentAction:", err);
    res.status(500).json({ error: "Errore eliminazione azione" });
  }
});

// ✅ Aggiorna stato/output azione
router.patch("/actions/:userId/:actionId", async (req, res) => {
  try {
    const { userId, actionId } = req.params;
    const { status, output } = req.body;
    const updates = {};
    if (status) updates.status = status;
    if (output) updates.output = output;

    await db
      .collection("ai_agent_hub")
      .doc(userId)
      .collection("actions")
      .doc(actionId)
      .update(updates);
    res.status(200).json({ message: "✅ Azione aggiornata" });
  } catch (err) {
    console.error("❌ updateAgentAction:", err);
    res.status(500).json({ error: "Errore aggiornamento azione" });
  }
});

// ✅ Recupera documenti IA
router.get("/documents/:userId", async (req, res) => {
  try {
    const { userId } = req.params;
    const snapshot = await db
      .collection("ai_agent_hub")
      .doc(userId)
      .collection("documents")
      .orderBy("createdAt", "desc")
      .get();
    const docs = snapshot.docs.map((doc) => doc.data());
    res.status(200).json(docs);
  } catch (err) {
    console.error("❌ getAgentDocuments:", err);
    res.status(500).json({ error: "Errore recupero documenti IA" });
  }
});

// ✅ Recupera configurazione IA
router.get("/config/:userId", async (req, res) => {
  try {
    const { userId } = req.params;
    const doc = await db.collection("ai_agent_hub_config").doc(userId).get();
    if (!doc.exists)
      return res
        .status(200)
        .json({ autonomyLevel: "base", enabledAutomations: {} });
    res.status(200).json(doc.data());
  } catch (err) {
    console.error("❌ getAgentConfig:", err);
    res.status(500).json({ error: "Errore recupero config" });
  }
});

// ✅ Salva configurazione IA
router.post("/config/:userId", async (req, res) => {
  try {
    const { userId } = req.params;
    const data = req.body;
    await db
      .collection("ai_agent_hub_config")
      .doc(userId)
      .set(data, { merge: true });
    res.status(200).json({ message: "✅ Config salvata" });
  } catch (err) {
    console.error("❌ saveAgentConfig:", err);
    res.status(500).json({ error: "Errore salvataggio config" });
  }
});

// ✅ Trigger: Dispatch IA (protetto + rate limit)
router.post("/dispatch", withRateLimit(10, 60_000), async (req, res) => {
  try {
    const userId = req.user.uid;
    const {
      taskType,
      model = "gpt-4",
      context = {},
      priority = "normal",
    } = req.body;

    if (!taskType)
      return res.status(400).json({ error: "taskType è obbligatorio" });

    await trackIAUsage({ userId, type: taskType, model });

    const now = new Date();
    const actionRef = db
      .collection("ai_agent_hub")
      .doc(userId)
      .collection("actions")
      .doc();

    const action = {
      actionId: actionRef.id,
      type: taskType,
      status: "pending",
      startedAt: now,
      context,
      output: null,
      priority,
    };

    await actionRef.set(action);

    await sendNotification({
      userId,
      title: `Nuova azione IA: ${taskType}`,
      description: `È stata avviata un’azione "${taskType}" per l’agente IA.`,
      type: "ai",
    });

    res
      .status(200)
      .json({ message: "✅ Azione IA salvata", actionId: actionRef.id });
  } catch (err) {
    console.error("❌ dispatchAgentAction:", err);
    res.status(500).json({ error: "Errore salvataggio azione IA" });
  }
});

// ✅ Trigger: Event Matcher IA
router.post("/event-matcher", withRateLimit(10, 60_000), async (req, res) => {
  try {
    const userId = req.user.uid;
    const { eventType, note = "", booking = {}, customer = {} } = req.body;
    if (!eventType)
      return res.status(400).json({ error: "eventType obbligatorio" });

    const matches = matchEvent({ eventType, note, booking, customer });
    if (matches.length === 0)
      return res
        .status(200)
        .json({ message: "Nessuna azione IA suggerita", actions: [] });

    const now = new Date();
    const batch = db.batch();
    const actions = [];

    for (const match of matches) {
      const actionRef = db
        .collection("ai_agent_hub")
        .doc(userId)
        .collection("actions")
        .doc();
      const action = {
        actionId: actionRef.id,
        type: match.type,
        status: "pending",
        startedAt: now,
        context: { eventType, note, bookingId: booking.id || null },
        output: { reason: match.reason },
        priority: match.priority || "normal",
      };
      batch.set(actionRef, action);
      actions.push(action);
    }

    await batch.commit();

    res
      .status(200)
      .json({ message: `✅ ${actions.length} azione(i) IA create`, actions });
  } catch (err) {
    console.error("❌ eventMatcher:", err);
    res.status(500).json({ error: "Errore analisi evento IA" });
  }
});

// ✅ Feedback IA
router.post("/feedback", withRateLimit(10, 60_000), async (req, res) => {
  try {
    const userId = req.user.uid;
    const { actionId, rating, comment = "" } = req.body;
    if (!actionId || !["up", "down"].includes(rating)) {
      return res.status(400).json({ error: "Dati feedback non validi" });
    }

    const now = new Date();
    await db
      .collection("ai_agent_hub")
      .doc(userId)
      .collection("feedback")
      .doc()
      .set({
        actionId,
        userId,
        rating,
        comment,
        timestamp: now,
        expireAt: new Date(now.getTime() + 365 * 24 * 60 * 60 * 1000),
      });

    res.status(200).json({ message: "✅ Feedback salvato" });
  } catch (err) {
    console.error("❌ sendAgentFeedback:", err);
    res.status(500).json({ error: "Errore salvataggio feedback" });
  }
});

// ✅ agent/checkin/send-welcome
router.post(
  "/checkin/send-welcome",
  withRateLimit(10, 60_000),
  async (req, res) => {
    try {
      const userId = req.user.uid;
      const { bookingId, clientName, checkinDate, checkoutDate } = req.body;

      if (!bookingId || !clientName || !checkinDate || !checkoutDate) {
        return res
          .status(400)
          .json({ error: "Dati mancanti per il check-in IA" });
      }

      const now = new Date();
      const actionRef = db
        .collection("ai_agent_hub")
        .doc(userId)
        .collection("actions")
        .doc();

      const clientMessage = `
👋 Ciao ${clientName}!

La tua registrazione per il soggiorno dal ${checkinDate} al ${checkoutDate} è stata ricevuta ✅

📌 Check-in disponibile dalle 15:00
📌 Check-out entro le 11:00

🔐 Il codice di accesso ti verrà fornito il giorno dell’arrivo.

✨ Servizi disponibili:
- Colazione + €9,90 al giorno
- Late check-out + €19,00 (fino alle 14:00)
- Noleggio e-bike e SPA convenzionata

Scrivici se vuoi prenotare uno di questi extra 💬

Grazie per aver scelto StayPro! 🏡
`;

      const action = {
        actionId: actionRef.id,
        type: "checkin",
        status: "completed",
        startedAt: now,
        context: { bookingId, clientName, checkinDate, checkoutDate },
        output: { clientMessage },
      };

      await actionRef.set(action);

      console.log(
        "📤 Check-in IA completato. Messaggio simulato al cliente:\n",
        clientMessage
      );

      return res.status(200).json({
        message: "✅ Check-in IA completato + messaggio generato",
        actionId: actionRef.id,
      });
    } catch (err) {
      console.error("❌ Errore /checkin/send-welcome:", err);
      return res
        .status(500)
        .json({ error: "Errore registrazione check-in IA" });
    }
  }
);

// ✅ agent/checkin/generate-pdf
router.post(
  "/checkin/generate-pdf",
  withRateLimit(10, 60_000),
  async (req, res) => {
    try {
      const userId = req.user.uid;
      const { bookingId, clientName, checkinDate, checkoutDate } = req.body;

      if (!bookingId || !clientName || !checkinDate || !checkoutDate) {
        return res
          .status(400)
          .json({ error: "Dati incompleti per generazione PDF" });
      }

      const now = new Date();
      const reportId = `checkin_${bookingId}`;

      const content = `
📝 CHECK-IN SUMMARY (Mock)
Cliente: ${clientName}
Booking ID: ${bookingId}
Check-in: ${checkinDate}
Check-out: ${checkoutDate}

Servizi suggeriti:
✔ Colazione
✔ Late check-out
✔ SPA e noleggio e-bike

Generato automaticamente da StayPro AI.
`;

      const tags = generateDocumentTags({
        type: "checkin",
        content,
        clientName,
      });

      const docRef = db
        .collection("ai_agent_hub")
        .doc(userId)
        .collection("documents")
        .doc(reportId);

      await docRef.set({
        reportId,
        title: `Riepilogo Check-in ${clientName}`,
        content,
        tags,
        createdAt: now,
      });

      console.log("📄 Documento riepilogo check-in generato:\n", content);

      return res.status(200).json({
        message: "✅ Documento riepilogo check-in mock salvato",
        reportId,
      });
    } catch (err) {
      console.error("❌ Errore /checkin/generate-pdf:", err);
      return res
        .status(500)
        .json({ error: "Errore generazione documento check-in" });
    }
  }
);

// ✅ notifications/ai
router.post(
  "/notifications/ai",
  withRateLimit(10, 60_000),
  async (req, res) => {
    try {
      const userId = req.user.uid;
      const { title, description } = req.body;

      if (!title || !description) {
        return res
          .status(400)
          .json({ error: "title e description sono obbligatori" });
      }

      const now = new Date();
      const notificationRef = db
        .collection("notifications")
        .doc(userId)
        .collection("list")
        .doc();

      const notification = {
        id: notificationRef.id,
        type: "ai",
        title,
        description,
        timestamp: now,
        read: false,
      };

      await notificationRef.set(notification);

      return res.status(200).json({
        message: "✅ Notifica AI salvata correttamente",
        notificationId: notification.id,
      });
    } catch (err) {
      console.error("❌ Errore /notifications/ai:", err);
      return res.status(500).json({
        error: "Errore durante il salvataggio della notifica AI",
      });
    }
  }
);

// ✅ agent/upsell/suggest
router.post(
  "/agent/upsell/suggest",
  withRateLimit(10, 60_000),
  async (req, res) => {
    try {
      const userId = req.user.uid;
      const { propertyId, bookingId, clientName } = req.body;

      if (!propertyId || !bookingId || !clientName) {
        return res.status(400).json({ error: "Dati obbligatori mancanti" });
      }

      const propSnap = await db.collection("properties").doc(propertyId).get();
      if (!propSnap.exists) {
        return res.status(404).json({ error: "Struttura non trovata" });
      }

      const propertyData = propSnap.data();
      const extraServices = propertyData.extraServices || [];
      const suggestedExtras = extraServices.map((s) => s.name || s);

      const now = new Date();
      const actionRef = db
        .collection("ai_agent_hub")
        .doc(userId)
        .collection("actions")
        .doc();

      const action = {
        actionId: actionRef.id,
        type: "upsell",
        status: "completed",
        startedAt: now,
        context: { bookingId, clientName, propertyId },
        output: { suggestedExtras },
      };

      await actionRef.set(action);

      await sendNotification({
        userId,
        title: "Suggerimento Up-Sell IA",
        description: `Sono stati suggeriti ${suggestedExtras.length} servizi extra per ${clientName}`,
        type: "ai",
      });

      return res.status(200).json({
        message: "✅ Upsell IA generato con successo",
        actionId: actionRef.id,
        suggestedExtras,
      });
    } catch (err) {
      console.error("❌ Errore /agent/upsell/suggest:", err);
      return res
        .status(500)
        .json({ error: "Errore durante suggerimento upsell" });
    }
  }
);

// ✅ agent/upsell/generate-pdf
router.post(
  "/agent/upsell/generate-pdf",
  withRateLimit(10, 60_000),
  async (req, res) => {
    try {
      const userId = req.user.uid;
      const { bookingId, clientName, suggestedExtras = [] } = req.body;

      if (!bookingId || !clientName || suggestedExtras.length === 0) {
        return res
          .status(400)
          .json({ error: "Dati incompleti per generazione PDF" });
      }

      const now = new Date();
      const reportId = `upsell_${bookingId}`;

      const extrasText = suggestedExtras.map((s) => `✔ ${s}`).join("\n");

      const content = `
🎯 UPSALE REPORT – Suggerimenti AI

Cliente: ${clientName}
Booking ID: ${bookingId}

Servizi consigliati:
${extrasText}

🔹 Offerte suggerite automaticamente da StayPro AI
🔹 Ottimizza i profitti con il marketing intelligente
`;

      const tags = generateDocumentTags({
        type: "upsell",
        content,
        clientName,
      });

      const docRef = db
        .collection("ai_agent_hub")
        .doc(userId)
        .collection("documents")
        .doc(reportId);

      await docRef.set({
        reportId,
        title: `Offerta Upsell per ${clientName}`,
        content,
        tags,
        createdAt: now,
      });

      console.log("📄 Documento PDF mock upsell creato:\n", content);

      return res.status(200).json({
        message: "✅ PDF upsell mock generato e salvato",
        reportId,
      });
    } catch (err) {
      console.error("❌ Errore /agent/upsell/generate-pdf:", err);
      return res
        .status(500)
        .json({ error: "Errore generazione documento upsell" });
    }
  }
);
module.exports = onRequest({ cors: true }, router);

File: backend/ai_backend/routes/agent_routes.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from uuid import uuid4
from datetime import datetime
from typing import Optional, List
import firebase_admin
from firebase_admin import firestore

router = APIRouter()

# ✅ Inizializza Firebase se non attivo
if not firebase_admin._apps:
    firebase_admin.initialize_app()
db = firestore.client()

# ✅ Modelli
class AgentActionRequest(BaseModel):
    user_id: str
    type: str  # es: "report", "pricing", "email", "automation"
    context: dict

class AgentActionUpdate(BaseModel):
    status: Optional[str] = None
    output: Optional[dict] = None

class CheckInMessageRequest(BaseModel):
    user_id: str
    booking_id: str
    guest_name: str
    email: str
    phone: Optional[str] = None
    checkin_date: str
    services_included: list
    services_extra: Optional[list] = []

# ✅ TRACK - Tracciamento nuova azione
@router.post("/agent/track-action")
async def track_agent_action(request: AgentActionRequest):
    try:
        now = datetime.utcnow()
        action_id = str(uuid4())

        doc_ref = db.collection("ai_agent_hub").document(request.user_id)
        actions_ref = doc_ref.collection("actions").document(action_id)

        actions_ref.set({
            "actionId": action_id,
            "type": request.type,
            "status": "pending",
            "startedAt": now,
            "context": request.context,
            "output": {},
        })

        doc_ref.set({
            "userId": request.user_id,
            "lastActive": now,
            "lastCompletedAction": None,
            "pendingActions": firestore.ArrayUnion([action_id])
        }, merge=True)

        return {"message": "✅ Azione IA tracciata", "actionId": action_id}

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore tracciamento azione: {str(e)}")

# 🔄 GET - Recupera tutte le azioni per un utente
@router.get("/agent/actions/{user_id}")
async def get_user_actions(user_id: str):
    try:
        actions_ref = db.collection("ai_agent_hub").document(user_id).collection("actions")
        docs = actions_ref.stream()
        return [doc.to_dict() for doc in docs]
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore recupero azioni: {str(e)}")

# 🔄 PATCH - Aggiorna una singola azione
@router.patch("/agent/actions/{user_id}/{action_id}")
async def update_agent_action(user_id: str, action_id: str, update: AgentActionUpdate):
    try:
        ref = db.collection("ai_agent_hub").document(user_id).collection("actions").document(action_id)
        updates = {}
        if update.status:
            updates["status"] = update.status
            if update.status == "completed":
                db.collection("ai_agent_hub").document(user_id).update({
                    "lastCompletedAction": action_id,
                    "pendingActions": firestore.ArrayRemove([action_id])
                })
        if update.output:
            updates["output"] = update.output
        ref.update(updates)
        return {"message": "✅ Azione aggiornata"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore aggiornamento azione: {str(e)}")

# 🔄 DELETE - Elimina una azione IA
@router.delete("/agent/actions/{user_id}/{action_id}")
async def delete_agent_action(user_id: str, action_id: str):
    try:
        db.collection("ai_agent_hub").document(user_id).collection("actions").document(action_id).delete()
        return {"message": "✅ Azione eliminata"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore eliminazione azione: {str(e)}")

# 🔄 GET - Recupera stato HUB IA
@router.get("/agent/hub-status/{user_id}")
async def get_agent_hub_status(user_id: str):
    try:
        doc = db.collection("ai_agent_hub").document(user_id).get()
        if not doc.exists:
            raise HTTPException(status_code=404, detail="Utente non trovato")
        return doc.to_dict()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore recupero stato HUB: {str(e)}")

# 🔄 GET - Recupera documenti generati IA
@router.get("/agent/documents/{user_id}")
async def get_generated_documents(user_id: str):
    try:
        docs_ref = db.collection("ai_agent_hub").document(user_id).collection("documents")
        docs = docs_ref.stream()
        return [doc.to_dict() for doc in docs]
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore recupero documenti: {str(e)}")

# 🔄 GET - Recupera configurazione agente (autonomia, automazioni, piano)
@router.get("/agent/config/{user_id}")
async def get_agent_config(user_id: str):
    try:
        doc_ref = db.collection("ai_agent_hub").document(user_id)
        doc = doc_ref.get()
        if not doc.exists:
            return {
                "autonomyLevel": "base",
                "enabledAutomations": {},
                "plan": "BASE"
            }
        data = doc.to_dict()
        return {
            "autonomyLevel": data.get("autonomyLevel", "base"),
            "enabledAutomations": data.get("enabledAutomations", {}),
            "plan": data.get("plan", "BASE")
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore recupero config: {str(e)}")

# ✅ POST - Invio messaggio di benvenuto check-in
@router.post("/agent/checkin/send-welcome")
async def send_checkin_welcome(request: CheckInMessageRequest):
    try:
        now = datetime.utcnow()
        message = f"""
Ciao {request.guest_name},

Benvenuto nella nostra struttura! 🎉
Ecco i dettagli della tua prenotazione:
- Check-in previsto: {request.checkin_date}
- Servizi inclusi: {', '.join(request.services_included)}
"""
        if request.services_extra:
            message += f"- Servizi extra disponibili: {', '.join(request.services_extra)}\n"

        message += "\n📩 Riceverai un'email o un messaggio WhatsApp con tutte le istruzioni utili.\nGrazie per aver scelto la nostra struttura!"

        action_id = str(uuid4())
        action_ref = db.collection("ai_agent_hub").document(request.user_id).collection("actions").document(action_id)
        action_ref.set({
            "actionId": action_id,
            "type": "checkin_welcome",
            "status": "completed",
            "startedAt": now,
            "context": {
                "bookingId": request.booking_id,
                "guestName": request.guest_name,
                "checkinDate": request.checkin_date,
                "servicesIncluded": request.services_included,
                "servicesExtra": request.services_extra
            },
            "output": {
                "message": message,
                "deliveredVia": "simulated"
            }
        })

        return {"message": "✅ Messaggio di benvenuto generato e salvato", "preview": message}

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore invio messaggio di benvenuto: {str(e)}")

# ✅ POST - Upload report da contesto IA (usato da agent/chat)
@router.post("/agent/upload-report")
async def generate_report_from_chat(context: dict):
    try:
        user_id = context.get("user_id", "unknown")
        session_id = context.get("session_id", "unknown")

        report_content = f"📊 Report generato per l’utente {user_id} – Sessione: {session_id}"
        timestamp = datetime.utcnow()
        doc_id = str(uuid4())

        doc_ref = db.collection("ai_agent_hub").document(user_id).collection("documents").document(doc_id)
        doc_ref.set({
            "documentId": doc_id,
            "type": "report",
            "content": report_content,
            "generatedAt": timestamp,
            "linkedSession": session_id,
        })

        return {
            "status": "completed",
            "message": "📎 Report generato con successo!",
            "documentId": doc_id,
            "content": report_content,
            "generatedAt": timestamp.isoformat()
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

# ✅ MODELLO PROFILO STRUTTURA
class Geo(BaseModel):
    lat: float
    lng: float

class StructureProfile(BaseModel):
    user_id: str
    name: str
    structureType: str
    style: Optional[str]
    description: Optional[str]
    address: str
    city: str
    country: str
    geo: Optional[Geo]
    checkin: Optional[str]
    checkout: Optional[str]
    rules: Optional[List[str]]
    services: Optional[List[str]]
    extraServices: Optional[List[str]]
    animalsAllowed: Optional[bool]
    languages: Optional[List[str]]
    maxGuests: Optional[int]
    rooms: Optional[int]
    preferredContact: Optional[str]
    phoneNumber: Optional[str]
    conventions: Optional[List[str]]
    transports: Optional[List[str]]
    shuttleService: Optional[bool]
    accessibility: Optional[str]
    notes: Optional[str]

# ✅ POST - Salva profilo struttura
@router.post("/agent/profile")
async def save_structure_profile(profile: StructureProfile):
    try:
        doc_ref = db.collection("ai_agent_hub").document(profile.user_id).collection("properties").document("main")
        doc_ref.set(profile.dict(exclude={"user_id"}), merge=True)
        return {"message": "✅ Profilo struttura salvato correttamente"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore salvataggio profilo: {str(e)}")

# ✅ GET - Recupera profilo struttura
@router.get("/agent/profile/{user_id}")
async def get_structure_profile(user_id: str):
    try:
        doc_ref = db.collection("ai_agent_hub").document(user_id).collection("properties").document("main")
        doc = doc_ref.get()
        if not doc.exists:
            raise HTTPException(status_code=404, detail="Profilo struttura non trovato")
        return doc.to_dict()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Errore recupero profilo: {str(e)}")